pkg/test/framework/resource/context.go

```diff
@@ -20,6 +20,11 @@ type Context interface {
 	// cleaned up.
 	TrackResource(r Resource) ID
 
+	// GetResource gets a resource matching the type referenced by the passed in pointer if one exists.
+	// Only useful when only one of the given type of resource is expected in the current context.
```

Comment:
could we not return all of a given type? Maybe the interface could be something like `GetResources(typ Resource) []Resource`?

Reply:
A couple usability things with that signature: 

1. Using the `Resource` interface requires the user to pass a concrete struct for `typ` in cases where the more specific interface doesn't embed `resource.Resource`. Ex: [resource.Cluster](https://github.com/istio/istio/blob/master/pkg/test/framework/resource/cluster.go#L25)
```go
var clusterTyp resource.Cluster // var declaration so we can pass an interface type
ctx.GetResources(clusterTyp)    // compile error
ctx.GetResource(kube.Cluster{}) // workaround, kube.Cluster implements resource.Resource

```

This might not be so bad.  It could encourage wrapping the ctx.GetResources call for individual resources. 
```go
func GetClusters(ctx resource.Context) []resource.Cluster {
    var res []resource.Resource
    ctx.Environment().case(environment.Kube, func() {
        res = ctx.GetResources(kube.Cluster{})
    })
    ctx.Environment().case(environment.Native, func() {
        res = ctx.GetResources(native.Cluster{})
    })
    // convert to []resource.Cluster and return
}
```

2. We'd need to do extra casting if we return the result instead of assigning: 
```go
clusterRes := ctx.GetResources(kube.Cluster{})
cluster := clusterRes[0].(kube.Cluster)
// or for the entire slice
clusters := make([]resource.Cluster, len(clusterRes))
for i, c := range clusterRes { /* cast and copy */ }
```

Reply:
1. That's fair. Let's stick with `interface{}`.

2. If we don't allow multiple values, I suspect we'll eventually run into scenarios where a caller needs a specific instance. Adding component-specific helper methods would help here. I imagine `Istio` will likely be the main component we'll be looking up anyway.

Reply:
Updated to allow multiple values by passing a slice instead of a pointer. This means we won't return an error if a resource is not found. The only error case is if the reference is not a pointer or slice. Maybe a panic makes sense? We'd be able to really clean up the signature to just `GetResources(ref interface{}`.

Another thought is that maybe it makes even more sense to split the plural case into a separate method.

Reply:
I think the separation between plural and singular could be done at the helper method level. See my suggestion for the Istio helper.